feat: InventoryPreset for InventoryManager

sqlerrorthing opened the pull request:
all new icons was taken from [svgrepo.com](https://www.svgrepo.com)

closes https://github.com/CCBlueX/LiquidBounce/issues/5937
closes https://github.com/CCBlueX/LiquidBounce/issues/3764

## Images
![image](https://github.com/user-attachments/assets/0d09afb4-d613-4146-9a78-70c4c2795978)

sqlerrorthing commented:
ui part was done (maybe)

sqlerrorthing commented:
i need help to implement [InventoryPresetsValue] to [ModuleInventoryCleaner]

sqlerrorthing commented:
I think I've implemented it, but I'm sure the old inv manager & chest stealer code needs to be cleaned up.

superblaubeere27 commented on src-theme/src/routes/clickgui/setting/inventoryPresets/ItemImage.svelte:
```diff
@@ -0,0 +1,33 @@
+{:else if item.type === "FOOD"}
+    <img src="img/clickgui/icon-food.svg" alt="Food"/>
+{:else if item.type === "BLOCKS"}
+    <img src="img/clickgui/icon-blocks.svg" alt="Blocks"/>
```
This icon does not look like a block. It looks like a copy icon.

sqlerrorthing replied:
in this context it looks like a block, I don't see any problem with this icon.

superblaubeere27 replied:
@1zun4 @SenkJu opinion: Does that icon represent `any block` for you in that context?

![image](https://github.com/user-attachments/assets/8c6ceac7-4b0a-4318-8f5c-7f072f0c487d)

sqlerrorthing replied:
I hope you understand that you can offer a better icon and I will agree (if it think so).

1zun4 replied:
> @1zun4 @SenkJu opinion: Does that icon represent `any block` for you in that context?
> 
> ![image](https://private-user-images.githubusercontent.com/22642852/429029318-8c6ceac7-4b0a-4318-8f5c-7f072f0c487d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1NDU3MzgsIm5iZiI6MTc0MzU0NTQzOCwicGF0aCI6Ii8yMjY0Mjg1Mi80MjkwMjkzMTgtOGM2Y2VhYzctNGIwYS00MzE4LThmNWMtN2YwNzJmMGM0ODdkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDAxVDIyMTAzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhNjc3MWQ2ZWI0MThhMmI0MjZlODVhN2JlMmVhOGUyOWU3ZDE0NWQ0Mjg3MTQ3NjhmNjZkMzRiYmMzYzJjNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RLk5HE_j4WFY8taLElCYz-378tUVVZIfMAMkjfSQiw0)

I could not come up with a better icon.

superblaubeere27 commented on src-theme/src/routes/clickgui/setting/inventoryPresets/config/PresetDeleteButton.svelte:
```diff
@@ -0,0 +1,57 @@
+        "Delete",
+        "Are you sure?",
+        "ARE YOU SURE?"
+    ]
```
Two clicks are enough.

superblaubeere27 commented on src-theme/src/routes/clickgui/setting/inventoryPresets/ItemImage.svelte:
```diff
@@ -0,0 +1,33 @@
+{:else if item.type === "BLOCKS"}
+    <img src="img/clickgui/icon-blocks.svg" alt="Blocks"/>
+{:else if item.type === "ANY"}
+    <img src="img/clickgui/icon-any.svg" alt="Any"/>
```
Use an asterisk instead

sqlerrorthing replied:
i dont like it

superblaubeere27 commented on src-theme/src/routes/clickgui/setting/inventoryPresets/ItemImage.svelte:
```diff
@@ -0,0 +1,33 @@
+{:else if item.type === "TOOLS"}
+    <img src="img/clickgui/icon-tools.svg" alt="Tools"/>
+{:else if item.type === "FOOD"}
+    <img src="img/clickgui/icon-food.svg" alt="Food"/>
```
Use something like an apple instead.

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/inventoryPresets/items/ToolsPresetItem.kt:
```diff
@@ -0,0 +1,27 @@
+    EnchantmentValueEstimator.WeightedEnchantment(Enchantments.FORTUNE, 0.33f),
+)
+
+data object ToolsPresetItem : PresetItem(
+    type = ItemType.TOOLS,
+    comparatorChain = ComparatorChain(
+        compareBy { (it.item as MiningToolItem).material().speed },
+        compareBy { VALUE_ESTIMATOR.estimateValue(it) },
+        PREFER_BETTER_DURABILITY,
+        STABILIZE_COMPARISON
+    )
+) {
+    override fun satisfies(stack: ItemStack) = stack.item is MiningToolItem
+}
```
Way to unspecific.

Mostly you don't want any tool, but you want something specific like a shovel.

sqlerrorthing replied:
I don't think that on pvp servers (I think they use inventory manager there much more often) anyone will really want this.

superblaubeere27 replied:
Of course they do. When playing bed wars you will specifically need an axe or a pickaxe for breaking the blocks surrounding the bed.

sqlerrorthing replied:
just put two slots of the tool type, like in the 9th and 8th slots a tool, in the 7th scissors, when there is a pickaxe and an axe they will take their places in the 9th or 8th slots, when there are scissors they will be in the 7th slot.

sqlerrorthing replied:
When breaking, [AutoTool] will do all the work for you.

superblaubeere27 commented on src-theme/src/routes/clickgui/setting/inventoryPresets/config/PresetModal.svelte:
```diff
@@ -0,0 +1,153 @@
+            />
+        </div>
+
+        <div class="throws-items-container">
+            <div class="header">
+                {throwsRendered}
+                <span class="muted">THROWS</span>
+                {#if throwsRendered > 0}
+                    <span class="clear-throws" on:click={handleClearAllThrows}>Clear all</span>
+                {/if}
+            </div>
+
+            <ThrowItemsContainer
+                    bind:this={throwsContainer}
+                    bind:throws={preset.throws}
+                    bind:renderedLength={throwsRendered}
+                    on:change={handleChange}
+            />
+        </div>
```
"Throws" is not a good name. I'd propose something like "max/min counts".

superblaubeere27 reviewed (changes requested):
Great. This will be a great and long awaited addition to the client!

## Frontend
The UI looks promising. I think there are still some problems with the current iteration of the design. In order to reduce configuration breaking changes we should also add other planned configuration features in this update. 

**Note that the following ideas are not specific instructions!** As long as you address the problems you can use your own designs and ideas.

What behavior should the user expect from this configuration?
![image](https://github.com/user-attachments/assets/f9de3fe1-b440-4650-a0e1-edb342d8cef7)
Why would the user add multiple presets at once? If you want the user to be able to between presets, then there should be an option to select an active preset.

The "throws" menu does not represent the real world complexity yet. The users often want to select how many items of a specific type they want to have or want to configure to throw out everything above a certain threshold.
![image](https://github.com/user-attachments/assets/744cb152-6e95-4d56-a753-c2d2f0906094)
This is my suggested UI. It is a list of rules. The higher the rule, the higher the priority (could be indicated by numbers). The 'I's are items. The slider on the left could be used to move rules around and change their priority. There can be multiple items in a single group. Since that might be ambiguous one could consider adding an option like group (all counts of all items in that group are compared against the limit) vs. each (the given constraints are applied per item). The slider on the right selects what number items should be kept at minimum and when it should start throwing items out, regardless of other rules. This may also just be a slider with a single value since both settings are pretty similar.

The item selection could also profit from some more customizability. Currently only one item or item type can be assigned to a specific item slot. There are common situations where you need more though. For example if you want to ignite another player you would probably have preferences like `lava bucket > flint and steel > firecharge`. Or sometimes some items have similar functions (like eggs or snowballs).
![asdfj](https://github.com/user-attachments/assets/6f9df6f4-b415-4a56-9b14-f97b09c5491f)
Therefore I propose this design. Again, there are prioritized rules (from top to bottom). If a higher rule cannot be satisfied, the lower will apply (for example this will try to fill the slot with an ender pearl, if that is not possible, it will try to fill it with a rod, eggs or snowballs. If that is again not possible, it will fill the slot with blocks). Additionally the user can select multiple specific items per group. If the user selects an item group, they cannot add additional items to that group though (for example you cannot have a rule that states "blocks or enderpearls"). Pressing "+" would bring the user to the next dialog.

The dialog to edit such a rule could look like this:
![ghjghj](https://github.com/user-attachments/assets/5fd972d2-f052-4137-8950-3753b84ba304)
I removed the "common items" category because I think it is distracting while offering not much benefit. I added radio boxes to show the user that they have to decide between selecting specific items and item categories.

As I said, those proposals are just suggestions.

## Backend
I think that some major changes to the current Inventory Cleaner backend will be necessary. And I also see that your implementation tries to address this and I know that you probably invested some time and effort into making this. Nevertheless, there are some severe issues:
1. The Inventory Cleaner now doesn't manage armor waste management anymore
2. The inventory cleaner is much less flexible now
3. The current item rating system is used in many modules. There is no point in having two seperate rating systems with the same code.
4. The current inv cleaner system has planning and execution separated. This is a feature which should be preserved.
5. The InventoryCleaner module had that name for quite some time now. Renaming it will confuse users even though there is an alias.
6. Never change a running system. The current backend is designed with something like this in mind. Only a few changes would be required to change this to support your and all of my proposed configuration changes.

I could do the necessary changes in the current implementation to make the configuration work if you want.

## Summary
The backend code will not make it into the final patch. The frontend code is good work. You proofed that you know how to create UIs, so I know that you will be able to bring this configuration to perfection ;D

sqlerrorthing commented:
# ~why can a user add multiple presets at once~

~because there is a merging functionality, the final decision on where each item will be in the inventory is taken into account based on the items that are in the inventory.~

### ~Example~
~_the more important the preset, the lower its index_~

~if the most important preset is in the first slot blocks
and in the second slot there is, for example, an apple
but at the time of sorting there is not a single block that can be put in the first slot, but there is an apple
then the apple will be put in the first slot.
that is why there is the possibility of several presets, to make the settings more flexible.~

---

~I think it's better to leave the system when the selection of the active preset will happen automatically based on the current items in the inventory directly. Perhaps, it can confuse new users the first time, but they should learn about it just once (for example, record a tutorial on how to make presets) and then they will have a lot of opportunities for customization.~

~in my opinion, it would be better to record a video on how to set up a preset, and if there are no presets in the settings, then add a button that will be a link to the video with the settings.~

sqlerrorthing commented:
> I removed the "common items" category because I think it is distracting while offering not much benefit. I added radio boxes to show the user that they have to decide between selecting specific items and item categories.

I don't like radio buttons, the names better reflect the essence and say choose one of two.

sqlerrorthing commented:
> I could do the necessary changes in the current implementation to make the configuration work if you want.

@superblaubeere27 I will try to take into account your wishes in the very setup and implementation of the frontend in particular, but yes, I need help in implementing the backend, I will be very grateful if you help me

sqlerrorthing commented:
_upd: The `Limit` replaced by `Max`_
_upd: The `Create new limit` replated by `Create a new limit`_

@superblaubeere27 i think i implemented what you wanted
_so far only on the frontend, and corrected the backend, these changes in inventoryCleaner are not implemented at all._

![image](https://github.com/user-attachments/assets/e03482a3-cd91-49b9-9620-a1d6de656eee)
in this case, if there are more than two stacks of blocks in the inventory, the extra ones will have to be thrown away
also, if there are more than 32 eggs OR snowballs (taken together, that is, their total amount in the inventory), then as many will be thrown out as are needed so that there are EITHER 32 eggs OR snowballs in total for the ENTIRE inventory.

but this requires implementation directly in the inventory manager & chest stealer itself
I can't do it because I don't understand how it works.

sqlerrorthing commented:
item selector now:
![image](https://github.com/user-attachments/assets/e681f58b-e975-456c-8abb-bc5cb800202b)

superblaubeere27 commented:
> # why can a user add multiple presets at once
> 
> because there is a merging functionality, the final decision on where each item will be in the inventory is taken into account based on the items that are in the inventory.

ok, so that is the same idea of allowing multiple choices for a single hotbar slot. I find this behavior quite unintuitive though. When I saw it for the first time and you asked me what multiple presets do then this behavior wouldn't have been in my first guesses.

Is that functionality equivalent with giving the user multiple choices per item?

> Perhaps, it can confuse new users the first time, but they should learn about it just once (for example, record a tutorial on how to make presets)

“A user interface is like a joke. If you have to explain it, it’s not that good”. — Martin Leblanc

“If you find an element of your interface requires instructions, then you need to redesign it.” — Dan Rubin



> I don't like radio buttons, the names better reflect the essence and say choose one of two.

That's ok as long as the user understands that they can only select either specific items or item categories.



> I need help in implementing the backend, I will be very grateful if you help me

I will implement every feature this PR needs from the `CleanupPlanGenerator` in another PR.



> @superblaubeere27 i think i implemented what you wanted _so far only on the frontend, and corrected the backend, these changes in inventoryCleaner are not implemented at all._
> 
> ![image](https://private-user-images.githubusercontent.com/148702857/428984857-e03482a3-cd91-49b9-9620-a1d6de656eee.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1MTIzNTUsIm5iZiI6MTc0MzUxMjA1NSwicGF0aCI6Ii8xNDg3MDI4NTcvNDI4OTg0ODU3LWUwMzQ4MmEzLWNkOTEtNDliOS05NjIwLWExZDZkZTY1NmVlZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQwMVQxMjU0MTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iY2VjODk1OTVhY2FlM2FmNjg2ZjVlMTg2OWRkYzNlOTQ5NDJiNjI5OGJkNzVmYjQ2MTFkNjEzMmJlNzA3ZWY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.26z808HLfoOb1CErQLlBv8kXXGoXM23G0QXcdfJXoUs) in this case, if there are more than two stacks of blocks in the inventory, the extra ones will have to be thrown away also, if there are more than 32 eggs OR snowballs (taken together, that is, their total amount in the inventory), then as many will be thrown out as are needed so that there are EITHER 32 eggs OR snowballs in total for the ENTIRE inventory.

Looks great 👍🏻

sqlerrorthing commented:
> I will implement every feature this PR needs from the CleanupPlanGenerator in another PR.

I'm looking forward to it.

# Please take this into account:
### Hotbar:
1. if `NonePresetItem` then there should be nothing in this slot
2. if `AnyPresetItem` then ignore the slot

### Limits:
if `AnyPresetItem` then all items will be taken into account and will be taken into account in calculations.

sqlerrorthing commented:
> Is that functionality equivalent with giving the user multiple choices per item?

maybe.. yes, but I have absolutely no idea how... it can be designed
you gave me your idea, but after looking at it I couldn't figure out how and where to put it, if you give me a bigger description of your idea, what it will look like, then I will listen to you

> “A user interface is like a joke. If you have to explain it, it’s not that good”. — Martin Leblanc
> “If you find an element of your interface requires instructions, then you need to redesign it.” — Dan Rubin

I am not a designer and will never position myself as a web developer. It literally gives me nothing.

sqlerrorthing commented:
> I am not a designer and will never position myself as a web developer. It literally gives me nothing.

also i'm from Russia, i don't even know them indirectly.
fuck putin. i also hate that i was born at this time, in this country, in this world

sqlerrorthing commented:
> ok, so that is the same idea of allowing multiple choices for a single hotbar slot. I find this behavior quite unintuitive though. When I saw it for the first time and you asked me what multiple presets do then this behavior wouldn't have been in my first guesses.

I thought up your idea and how to implement it, yes, it is still worth abandoning the system with multiple presets (which are dynamically combined at the sorting stage). simply adding the ability to select several candidates for a particular slot.

sqlerrorthing commented:
_upd: my implementation in InventoryCleaner was also removed (but I didn't change the name back)_

![image](https://github.com/user-attachments/assets/ff0557e5-c198-43c1-a02c-8382fe0d01cd)
![image](https://github.com/user-attachments/assets/5a866c63-be66-44af-aa31-a9ade9f9a805)
https://github.com/sqlerrorthing/LiquidBounce/pull/18

#### Candidates priority:
`a` - above, `d` - below

```
a b
c d
```

---

the number in slot 4 (in the candidate preview) is limited to 9, i.e. `+9` is the maximum that can be displayed in the preview (otherwise the ui may just leave.)

sqlerrorthing commented:
I think the frontend part is completely ready, waiting for news from @superblaubeere27.
~for now I'll document the code in the backend~ (documented)

superblaubeere27 commented:
I have implemented it on a local branch. I will merge it soon.

sqlerrorthing commented:
> I have implemented it on a local branch. I will merge it soon.

good 👍

sqlerrorthing commented:
@superblaubeere27 well how are things there?

superblaubeere27 commented:
I pushed my changes. I don't know why it does not work yet. Can you try to debug the frontend?

sqlerrorthing commented:
> I pushed my changes. I don't know why it does not work yet. Can you try to debug the frontend?

maybe okay. but. please, revert yr changes & create pull request on my repo

superblaubeere27 commented:
Why?

sqlerrorthing commented:
well then if it already works then I think it's not necessary anymore.

sorry, but I don't have time for this at all, next weekend I'll do it again.

sqlerrorthing commented:
i dont like these icons on commit https://github.com/CCBlueX/LiquidBounce/pull/5938/commits/2adae944c76c2abcc1b6fd9260714c978ba85063

superblaubeere27 commented:
> i dont like these icons on commit [2adae94](https://github.com/CCBlueX/LiquidBounce/commit/2adae944c76c2abcc1b6fd9260714c978ba85063)

For context, these icons:

![image](https://github.com/user-attachments/assets/c1b1eb4d-2d7a-4282-8093-77a00c4f62d0)

superblaubeere27 commented:
This PR is almost done.

There are still some points we need to address:
1. Decide which items should be in the quick select section (12 items total)
2. Decide what the default config should be
3. Agree on icons
4. Move the `Any` and `Ignore` item option to `PresetItemGroupCandidateSelector.svelte`
5. Add tooltips to clarify stuff like `Any` and `Ignore` for the user.
6. Fix remaining bugs

## Bugs
- Setting a specific number on the item limit slider does not work yet

superblaubeere27 commented:
The backend is finished.

sqlerrorthing commented:
> sorry, but I don't have time for this at all, next weekend I'll do it.

sqlerrorthing commented:
> The backend is finished.

> Setting a specific number on the item limit slider does not work yet

still as bug?

sqlerrorthing commented:
sword, blocks, and etc useful items, e.g road, lava/water bucket, thorwable n etc

sqlerrorthing commented on build.gradle:
```diff
@@ -32,6 +32,10 @@ def getContributors(repoOwner, repoName) {
 
     connection.setRequestProperty("User-Agent", "LiquidBounce-App")
 
+    if (true) {
+        return []
+    }
+
```
???

superblaubeere27 replied:
The gradle build didn't work for me for whatever reason. So I eliminated the problematic section. We can just revert this change.

superblaubeere27 commented:
> > The backend is finished.
> 
> > Setting a specific number on the item limit slider does not work yet
> 
> still as bug?

The bug is that currently you cannot type in the item count you want, the frontend won't do it. The backend seems to be good though.

1zun4 commented:
A default configuration and presets would be great. I tested the PR and it worked for me, but I think @SenkJu needs to tweak the editor a bit as it's visually very inconsistent with the rest of the UI.

sqlerrorthing commented:
> A default configuration and presets would be great.

maybe we can make a default configuration based on skywars? in my opinion, this is what they use most often

sqlerrorthing commented:
> Add tooltips to clarify stuff like Any and Ignore for the user.

added [here](https://github.com/sqlerrorthing/LiquidBounce/blob/1669781993f6366ff836077ac320bb55738fd57a/src-theme/src/routes/clickgui/setting/inventoryPreset/config/PresetItemSelector.svelte#L111) n [here](https://github.com/sqlerrorthing/LiquidBounce/blob/1669781993f6366ff836077ac320bb55738fd57a/src-theme/src/routes/clickgui/setting/inventoryPreset/config/PresetItemSelector.svelte#L116)

sqlerrorthing commented:
I strongly disagree with the icons that are there now, they don't look very good, especially the sword - is it stone/iron? And the blocks? Are they some kind of dull wood? It's not intuitive at all, yes, there are hints, but it still looks terrible, even if it is understandable (because of the hints I added).

also, this does not fit at all with the fact that there is `Any` and `None`, everything should be in the same style.

there is also a thing as texture packs, and I am 100% sure that someone has the same colors for some objects.

sqlerrorthing commented:
bug:
![image](https://github.com/user-attachments/assets/f067bc73-f207-4ea7-a6ac-f644eb1831dc)
candidates: egg, snowball ignore

expected: if there are eggs - put eggs, if there are snowballs - put them, if not - the slot must be empty.

reality: the slot is always empty.
![image](https://github.com/user-attachments/assets/42815b51-7ce7-4020-a533-caaad4ec1b45)

sqlerrorthing commented:
https://github.com/sqlerrorthing/LiquidBounce/pull/21

> ## ![image](https://private-user-images.githubusercontent.com/148702857/433018202-c4471a4b-b15f-4023-b951-a622bdd68ee1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0NDIwNjYsIm5iZiI6MTc0NDQ0MTc2NiwicGF0aCI6Ii8xNDg3MDI4NTcvNDMzMDE4MjAyLWM0NDcxYTRiLWIxNWYtNDAyMy1iOTUxLWE2MjJiZGQ2OGVlMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxMlQwNzA5MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01OGUxY2RiY2I1ZDYwMGJmZjNmMGRjODI1MTE3NmY2M2VkNmVhNTJiYjlhZWUxZGIxOTZhMjZiNzYxY2NiYWY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4RQyTbIxDQyR3gXLq7lNfnteSMC1zUR6Un37PXOqBZ8)
> 
> ![image](https://private-user-images.githubusercontent.com/148702857/433018208-e3fa797d-eb22-457e-84c4-11edd1fc02b5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0NDIwNjYsIm5iZiI6MTc0NDQ0MTc2NiwicGF0aCI6Ii8xNDg3MDI4NTcvNDMzMDE4MjA4LWUzZmE3OTdkLWViMjItNDU3ZS04NGM0LTExZWRkMWZjMDJiNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxMlQwNzA5MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNjFkMjI0ZTc3M2I5MzFkNTFiOTgxOWQwNTQxMjE1MmI3ZmM5MDdhODU0ZDlmZDJhYmZkZjg5MDQ4MmUxNGVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.edthKhQHQa9nfY-igDINwNBIlXpUA2u_oDjJ7d_UGRE) ![image](https://private-user-images.githubusercontent.com/148702857/433018211-fdfd9b54-40eb-4ecf-8b55-a6f38a63d257.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0NDIwNjYsIm5iZiI6MTc0NDQ0MTc2NiwicGF0aCI6Ii8xNDg3MDI4NTcvNDMzMDE4MjExLWZkZmQ5YjU0LTQwZWItNGVjZi04YjU1LWE2ZjM4YTYzZDI1Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxMlQwNzA5MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YmE0MjMzZWM3M2MxYmRjNTFkYjg4ZGJmZWYyZmJlMjdjMDE2YWIxYjQ5ODY3ZmM0YjViMzEwNWM4NGE1NDViJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.rrfl9YnQSE3yl2bglp1x6rWgQIMQcsOmC5OLtWs5fAs) ![image](https://private-user-images.githubusercontent.com/148702857/433018216-eb3d0b83-8cf9-4749-902f-b45cbd3bcab3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0NDIwNjYsIm5iZiI6MTc0NDQ0MTc2NiwicGF0aCI6Ii8xNDg3MDI4NTcvNDMzMDE4MjE2LWViM2QwYjgzLThjZjktNDc0OS05MDJmLWI0NWNiZDNiY2FiMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxMlQwNzA5MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03ZGFjZDE2NDUxN2MxOTk0ZWJhZjdlN2U3YjBmYzk1OTNlNjZlMzFkNjg1N2E1Mzk1ZmIxOGRlNjU2YmJmZmQxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.CpRp0MoH4ILEA2JEVU39xhw3S-wTacR62k1ye02Z8d8) _items change cyclically once every 1 second_

sqlerrorthing commented:
https://github.com/sqlerrorthing/LiquidBounce/pull/22
> ![image](https://private-user-images.githubusercontent.com/148702857/433042960-487708a5-2faa-4b0c-8d88-f3114484dc85.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0NTg2MDEsIm5iZiI6MTc0NDQ1ODMwMSwicGF0aCI6Ii8xNDg3MDI4NTcvNDMzMDQyOTYwLTQ4NzcwOGE1LTJmYWEtNGIwYy04ZDg4LWYzMTE0NDg0ZGM4NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxMlQxMTQ1MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MGY4ZTMyZTIyMzVhYzdlMDc3YmY3NGRjNTYxYTFiMmFlMWZiN2E1YzVjNGFlZGI2NjBjZWY2ZDY5NGM4M2E0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Fh6U8GIFlFhZEkIJQfvZW8eObfzO4W4BG3OwxIg2mqw)

sqlerrorthing commented:
![image](https://github.com/user-attachments/assets/7f4697a6-20cb-4022-99d9-7effc401dfe1)

bug: any item doesnt throw at least one item
![image](https://github.com/user-attachments/assets/06d7f297-7204-470d-9b03-15d028d4a6b7)
I expected that everything would be thrown out.

sqlerrorthing commented:
> > > The backend is finished.
> > 
> > 
> > > Setting a specific number on the item limit slider does not work yet
> > 
> > 
> > still as bug?
> 
> The bug is that currently you cannot type in the item count you want, the frontend won't do it. The backend seems to be good though.

maybe fixed in https://github.com/CCBlueX/LiquidBounce/pull/5938/commits/f0b6d871bfbf61d3b68c28f680d78690324d4c09

superblaubeere27 commented on src-theme/src/routes/clickgui/setting/inventoryPreset/config/PresetItemSelector.svelte:
```diff
@@ -0,0 +1,312 @@
+    export let setItem: (item: PresetItem) => void
+    export let filter: ((item: string) => boolean) | null = null
+
+    export let canSelectAny: boolean
+    export let canSelectIgnore: boolean
```
I think this is a bad way to do this. Any and ignore make no sense on item selection, because they mean 'no item'. 

A better way would be to add 3 options for a slot: (1) an item (current item selection), (2) any and (3) ignore

sqlerrorthing replied:
what do you mean?

superblaubeere27 commented on src/main/kotlin/net/ccbluex/liquidbounce/features/inventoryPreset/InventoryPreset.kt:
```diff
@@ -0,0 +1,60 @@
+        this.items = itemMap
+    }
+
+    private fun getSlotForIndex(idx: Int) = when (idx) {
+        0 -> OffHandSlot
+        else -> HotbarItemSlot(idx - 1)
+    }
```
Please don't use multiline expression functions.

superblaubeere27 commented on src-theme/src/routes/clickgui/setting/inventoryPreset/config/PresetItemSelector.svelte:
```diff
@@ -0,0 +1,312 @@
+        ...(canSelectAny ? [{
+            item: { type: "ANY" } as const,
+            name: "Any",
+            tooltip: "This slot is simply skipped during filtering..."
```
Any just says that you have no content preference for this slot

superblaubeere27 commented on src-theme/src/routes/clickgui/setting/inventoryPreset/config/PresetItemSelector.svelte:
```diff
@@ -0,0 +1,312 @@
+        ...(canSelectIgnore ? [{
+            item: { type: "IGNORE" } as const,
+            name: "Ignore",
+            tooltip: "There can be nothing in this slot"
```
Ignore means that the InvCleaner may not touch anything which is in this slot (even if it considers it useless)

superblaubeere27 commented:
> I strongly disagree with the icons that are there now, they don't look very good, especially the sword - is it stone/iron? And the blocks? Are they some kind of dull wood? It's not intuitive at all, yes, there are hints, but it still looks terrible, even if it is understandable (because of the hints I added).

We should probably add a hint like a small blue circle in the lower right corner to indicate that those are item groups.

superblaubeere27 commented:
> expected: if there are eggs - put eggs, if there are snowballs - put them, if not - the slot must be empty.

no, that is not the expected behavior at all! `Ignore` is something very different than `empty`. It just states that the slot's contents may not be touched.



> ![image](https://private-user-images.githubusercontent.com/148702857/433048475-7f4697a6-20cb-4022-99d9-7effc401dfe1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0OTMyMjYsIm5iZiI6MTc0NDQ5MjkyNiwicGF0aCI6Ii8xNDg3MDI4NTcvNDMzMDQ4NDc1LTdmNDY5N2E2LTIwY2ItNDAyMi05OWQ5LTdlZmZjNDAxZGZlMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxMlQyMTIyMDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNzAyMjUzY2VmNWQyMzIxNGQxYzVjN2E3NjIyYTMyNjg5MTBkYzIxYzQ0Y2U4NWQ3Zjk2ZmUyMzg3MTQ4NmZiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2zhpPdNOWFt5IzKqkZ6t0cKA2znCM_2TNUZ3S3HSylE)
> 
> bug: any item doesnt throw at least one item ![image](https://private-user-images.githubusercontent.com/148702857/433048490-06d7f297-7204-470d-9b03-15d028d4a6b7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ0OTMyMjYsIm5iZiI6MTc0NDQ5MjkyNiwicGF0aCI6Ii8xNDg3MDI4NTcvNDMzMDQ4NDkwLTA2ZDdmMjk3LTcyMDQtNDcwZC05YjAzLTE1ZDAyOGQ0YTZiNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxMlQyMTIyMDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iMzliYWQ4NmMwZWY4NTIxMmRkZjBlZTgwOGU2NWM2OTFmNzFiMjRmZDczZTU3MGIwMmZlNzZhMDIzYzcyM2JjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.IQxKv-nxspYZP950EJH3MEd6NaTxKMKUSknCz5ggdKg) I expected that everything would be thrown out.

with this you set the count of 'no special preference for this slot' to 0. That makes no sense!

Why would there be a setting which allows the user to set an absolute amount of items in their inventory in the first place. That would be really useless.

sqlerrorthing commented:
> no, that is not the expected behavior at all! Ignore is something very different than empty. It just states that the slot's contents may not be touched.

oh, my bad

sqlerrorthing commented:
> We should probably add a hint like a small blue circle in the lower right corner to indicate that those are item groups.

no, i think we should abandon the minecraft style group icons, i have compiled a new list of icons in the pr i indicated above, i am waiting for your opinions

sqlerrorthing commented:
> We should probably add a hint like a small blue circle in the lower right corner to indicate that those are item groups.

no, i think we should abandon the minecraft style group icons, i have compiled a new list of icons in the pr i indicated above, i am waiting for your opinions

sqlerrorthing commented:
_wrong commit (https://github.com/CCBlueX/LiquidBounce/pull/5938/commits/f27bc6ed47f91f954ae54ed7f6c0355969b72470) name_

add muted effect for unhovered items

MukjepScarlet commented on src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/invcleaner/ItemNumberConstraintEnforcer.kt:
```diff
@@ -0,0 +1,64 @@
+        constraints.sortBy { it.group.priority }
+
+        for (constraintInfo in constraints) {
+            val currentCount = this.currentLimit[constraintInfo.group] ?: 0
```
getOrDefault?

sqlerrorthing replied:
absolutly same thing.
this is doesnt make any sense and the kotlin style (`?:`) was used everyhere in this project

superblaubeere27 commented:
![image](https://github.com/user-attachments/assets/dfd30850-9ab7-48e2-916f-05ef746defbe)

sqlerrorthing commented:
> ![image](https://github.com/user-attachments/assets/dfd30850-9ab7-48e2-916f-05ef746defbe)
> 

![IMG_20250414_130554.jpg](https://github.com/user-attachments/assets/f9905406-7f40-4f19-9360-22640f077fd8)

?

sqlerrorthing commented:
> ![image](https://github.com/user-attachments/assets/dfd30850-9ab7-48e2-916f-05ef746defbe)
> 

![IMG_20250414_130554.jpg](https://github.com/user-attachments/assets/f9905406-7f40-4f19-9360-22640f077fd8)

?

![IMG_20250414_131410.jpg](https://github.com/user-attachments/assets/934a3322-b8db-4d3b-a90e-330db6f5a53e)

?

MukjepScarlet commented:
I suggest that use mono font like consolas for item counts

sqlerrorthing commented:
https://github.com/sqlerrorthing/LiquidBounce/pull/24
> ![image](https://github.com/user-attachments/assets/71ec4cb6-217e-45cf-8fa9-474591a36883)
> ![image](https://github.com/user-attachments/assets/b0df0466-49d1-443f-b7f6-736ebacd4a8d)

superblaubeere27 commented:
> [sqlerrorthing#24](https://github.com/sqlerrorthing/LiquidBounce/pull/24)
> 
> > ![image](https://private-user-images.githubusercontent.com/148702857/433367828-71ec4cb6-217e-45cf-8fa9-474591a36883.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2NjM1OTUsIm5iZiI6MTc0NDY2MzI5NSwicGF0aCI6Ii8xNDg3MDI4NTcvNDMzMzY3ODI4LTcxZWM0Y2I2LTIxN2UtNDVjZi04ZmE5LTQ3NDU5MWEzNjg4My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxNFQyMDQxMzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZTMwOTkxYmUwNTVlMmNkODU1MmQ3NDM1ZmQwZTRmZThlODE1MWRlOWU5NzQzMGY4ZmU2MGY5YzYwYzE5ZjQ4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7nHZEcgM1LmAJiM6xAntHvBvhkaJR5JnIptO_UkbO_I)
> > ![image](https://private-user-images.githubusercontent.com/148702857/433367920-b0df0466-49d1-443f-b7f6-736ebacd4a8d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2NjM1OTUsIm5iZiI6MTc0NDY2MzI5NSwicGF0aCI6Ii8xNDg3MDI4NTcvNDMzMzY3OTIwLWIwZGYwNDY2LTQ5ZDEtNDQzZi1iN2Y2LTczNmViYWNkNGE4ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxNFQyMDQxMzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zNzNjNWVjMDNiNzk0ODUwZDQyNDE3NDk5NGI1YzdhOWFiMzBmNTlhZjUxMjJkMjE4NTQzNTZiNjdlZWM3MjIzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.BmrCQUZCHuPacGnt9oEwd2LNVPMNBip6RNuNnjcWHXg)

I like it 👍🏻

sqlerrorthing commented:
alright, what should we do next? or wait for a response from @SenkJu?

sqlerrorthing commented:
guys, please, rate these icons from 0 up to 10
https://github.com/sqlerrorthing/LiquidBounce/pull/22
> ![image](https://github.com/user-attachments/assets/487708a5-2faa-4b0c-8d88-f3114484dc85)

1zun4 commented:
> guys, please, rate these icons from 0 up to 10 [sqlerrorthing#22](https://github.com/sqlerrorthing/LiquidBounce/pull/22)
> 
> > ![image](https://private-user-images.githubusercontent.com/148702857/433042960-487708a5-2faa-4b0c-8d88-f3114484dc85.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2OTY3MDIsIm5iZiI6MTc0NDY5NjQwMiwicGF0aCI6Ii8xNDg3MDI4NTcvNDMzMDQyOTYwLTQ4NzcwOGE1LTJmYWEtNGIwYy04ZDg4LWYzMTE0NDg0ZGM4NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxNVQwNTUzMjJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNzZjZGY2MDhkMTMxZTg5ZTA0YWE5ZDk5MjVhMGRmZDdiMTlhNWE0MGM2MWIzOGViNmFmMmY4MTI1OTZjNWY5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.9m85WlakXf983IGoHvqwrvL_itokxhwyeaSM7DetXJI)

They look very cartoon-like.

sqlerrorthing commented:
> They look very cartoon-like.

however, they correspond (I think) to the general style of icon design in the client

superblaubeere27 commented:
> guys, please, rate these icons from 0 up to 10 [sqlerrorthing#22](https://github.com/sqlerrorthing/LiquidBounce/pull/22)
> 
> > ![image](https://private-user-images.githubusercontent.com/148702857/433042960-487708a5-2faa-4b0c-8d88-f3114484dc85.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ3NDYwODcsIm5iZiI6MTc0NDc0NTc4NywicGF0aCI6Ii8xNDg3MDI4NTcvNDMzMDQyOTYwLTQ4NzcwOGE1LTJmYWEtNGIwYy04ZDg4LWYzMTE0NDg0ZGM4NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxNVQxOTM2MjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNzkxY2Q3NjZlN2U1OGY5MjBmNzE0NjM0ODY0OTg5YWNkNjk0ZWJhZTk5MmI0YWE3MzFjNzgxMGM1ODdmZjdkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.3kG59-1I0xIXVxMI5qSna8Xcy_IcpAbtfc6Qy8AhHPo)

they will cause a weird cartoon- and minecraft-icon mixup (5/10)

slosacoder commented:
> guys, please, rate these icons from 0 up to 10
> https://github.com/sqlerrorthing/LiquidBounce/pull/22
> > ![image](https://github.com/user-attachments/assets/487708a5-2faa-4b0c-8d88-f3114484dc85)
> 

what about using minecraft textures like sb27 did and just make them white instead of monochrome?

sqlerrorthing commented:
> what about using minecraft textures like sb27 did and just make them white instead of monochrome? (wondering how that would look like)

just do it by urself

sqlerrorthing commented:
![IMG_20250416_123116.jpg](https://github.com/user-attachments/assets/0ed4a156-5a9a-43d6-a31c-6ea5d220eb04)

ghost commented:
Crazy

sqlerrorthing commented:
> Crazy

but good...

TipoRick commented:
гениально

sqlerrorthing commented:
whats new?

superblaubeere27 commented:
We are currently waiting for a review/edit by @SenkJu.

sqlerrorthing commented:
> We are currently waiting for a review/edit by @SenkJu.

what @1zun4 think abt it?

1zun4 commented:
I already approved.

superblaubeere27 reviewed (commented):
# What needs to happen
We need to address:
1. [ ] Decide what the default config should be
2. [ ] Decide which items should be in the quick select section (12 items total) 

These issues need to be addressed by @SenkJu:
1. [ ] Agree on icons / add some indicator to the generic item icons (so you cannot confuse an iron sword with a generic sword)

sqlerrorthing commented:
> Decide which items should be in the quick select section (12 items total)

first - in pvp usage pvp (sword, bucket, lava bucket)
second - tools (axe, pickaxe, etc...)
third- throwables (eggs, snows or g.i. `Throwables`)

> Decide what the default config should be

for hypixel skywars maybe

---

maybe move these steps for description?

sqlerrorthing commented:
@SenkJu, whats up?

SenkJu commented:
Sorry, I have not forgotten about this pull request. I am unfortunately still very busy and reviewing and possibly refactoring 3000 lines of code will take a while. I will try to get to it as soon as possible.

sqlerrorthing commented:
@SenkJu sorry for the ping, but r u still very busy?

SenkJu commented:
I have started working on it. Please refrain from pushing any further adjustments until I am done refactoring.

sqlerrorthing commented:
whats new?

ghost commented:
@senkju left us 😔

superblaubeere27 commented:
😞✊🏻

sqlerrorthing commented:
is @SenkJu died?

SuicidalEeper commented:
> is @SenkJu died?

He does that a lot

1zun4 commented:
No, this is still planned to be worked on by @SenkJu. While the Kotlin side looks fine to me, the Svelte side does not seem to be up to standard, and @SenkJu has stated that the Svelte code needs to be rewritten. He is working on it, but he is taking his time to do it properly - he just doesn't communicate much on GitHub, preferring to do so on our private Discord server.

MukjepScarlet commented on src-theme/src/integration/host.ts:
```diff
@@ -1,4 +1,4 @@
-const IN_DEV = false;
+const IN_DEV = true;
```
hey...

sqlerrorthing replied:
🤫🤫🤫🤫🫢

MukjepScarlet commented:
Conflicts
see https://github.com/CCBlueX/LiquidBounce/pull/6854 https://github.com/CCBlueX/LiquidBounce/pull/6852
1. Remove usage of class SwordItem/MiningToolItem/AxeItem and all other tools
2. Remove usage of class ArmorItem
3. Remove usage of `player.inventory.armor`. Use `Slots.Armor` for the client player

MukjepScarlet commented:
https://github.com/CCBlueX/LiquidBounce/blob/9471bf6c48479db84b20cb78ccba8f39863c6fcc/src/main/kotlin/net/ccbluex/liquidbounce/utils/item/ItemCategorizations.kt#L60-L98

MukjepScarlet commented:
Who wanna continue this?

sqlerrorthing commented:
no :sob:
1. idk how fix backend
2. where is the issues that @SenkJu told need to be resolved

sqlerrorthing commented:
> @MukjepScarlet added the Theme label on Sep 26

there is no only theme changes

MukjepScarlet commented:
> idk how fix backend

Start a new branch after https://github.com/CCBlueX/LiquidBounce/pull/7402

Move the changed files to the new branch. And split them into multiple PRs if possible.

MukjepScarlet commented on src/main/kotlin/net/ccbluex/liquidbounce/utils/sorting/ComparatorChain.kt:
```diff
@@ -35,14 +35,10 @@ class ComparatorChain<T>(private vararg val comparisonFunctions: Comparator<in T
 }
 
 inline fun <T> compareValueByCondition(a: T, b: T, cond: (T) -> Boolean): Int {
```
Also, skip unrelated stuffs like this

MukjepScarlet commented:
I decide to continue this (probably)

MukjepScarlet committed:
squashed commit

MukjepScarlet committed:
revert

MukjepScarlet committed:
[Skip CI] revert

sqlerrorthing commented:
thank you cuz i wanna to close it

MukjepScarlet commented:
Maybe I need to cherrypick or redesign, too many conflicts

MukjepScarlet committed:
[ci skip] Update ModuleZoom.kt

MukjepScarlet committed:
[ci skip] Update ModuleElytraSwap.kt

superblaubeere27 committed:
WIP: train data collector and ML inference scaffolding

superblaubeere27 committed:
Merge pull request #5938 into nextgen: Inventory preset + invcleaner rewrite

Keep the PR's modern invcleaner architecture (WishOrganizer, ItemDispenserRack,
ItemNumberConstraintEnforcer, CleanupPlanTemplate, per-slot multi-wish
preferences, item limit/groups, inventory preset) and adapt it to nextgen's
platform (Mojang mappings, data-driven tools/armor via ItemStack tags,
nextgen's ItemSlot/Armor/Config system).

- Adapted the PR's ItemCategorization/facets to nextgen's tag-based extension
  functions (isSword/isMiningTool/isPlayerArmor/isAxe/...) replacing the
  removed SwordItem/MiningToolItem/ArmorItem/PickaxeItem classes.
- Wired the inventory preset value into nextgen's config system
  (ValueType.INVENTORY_PRESET, ValueGroup.inventoryPreset(), gson adapters).
- Ported nextgen consumers (ModuleChestStealer, SpeedGeneric, ModuleItemTags,
  debugrecorder) onto the new API.
- Integrated the PR frontend (types.ts, GenericSetting routing, svelte preset
  components) into nextgen's theme.

Verified: compileKotlin passes; theme vite build succeeds.

superblaubeere27 committed:
Restore mace/spear/shield/rod/arrow facets on the merged invcleaner

The PR's invcleaner taxonomy lacked dedicated types for these items, which
dropped their enchantment-aware ranking. Re-add them onto the merged
architecture so they are kept and sorted by their enchantments:

- Add GenericItemType members MACE, SPEAR, ROD, SHIELD.
- Restore ArrowItemFacet, RodItemFacet, ShieldItemFacet, MaceItemFacet,
  SpearItemFacet adapted to the PR ItemCategory model (Mace/Spear remain
  WeaponItemFacet specializations with DENSITY/BREACH/WIND_BURST and
  LUNGE/speed ranking).
- Route them in ItemCategorization.getItemFacets (FishingRodItem, ShieldItem,
  isSpear, MaceItem).
- Add PREFER_ENCHANTABLE comparator and make ItemFacet implement
  ItemStackHolder (by delegation) so holder comparators resolve.
- SpeedGeneric now uses SpearItemFacet.COMPARING_LUNGE_AND_SPEED (upstream).

compileKotlin passes.

superblaubeere27 committed:
Expose new invcleaner item facets in the preset UI

Make the restored SPEAR/MACE/SHIELD/ROD facets selectable as slot group
preferences in the inventory preset ClickGUI:

- Add SPEAR/MACE/SHIELD/ROD to the backend ItemGroupType enum
  (FrontendSlotPreference) mapping to the new GenericItemType members.
- Add them to the frontend GroupItemPreference group union in types.ts.
- Add matching group entries in the "Item Groups" quick select list.
- Add icon mapping + new SVG icons (spear/mace/shield/rod).

compileKotlin and the theme vite build both pass; no new svelte-check
errors beyond nextgen's pre-existing ones.

superblaubeere27 committed:
Remove ML/train-data-collector scaffolding from nextgen (kept on feat/train-data-collector)

Commit 06659b292 "WIP: train data collector and ML inference scaffolding"
did not belong in the invcleaner merge branch. It has been preserved in
its entirety on the `feat/train-data-collector` branch (same commit), and
this removes/reverts it from `nextgen`:

- Delete ModuleTrainDataCollector and the traindata package
  (DataSample, InferenceUtility, ItemCategoryProvider, POIProvider,
  ParquetWriterUtil, SurroundingsMapper).
- Delete associated docs, scripts, config (immediatelyfast/lithium/
  sodium) and test-parquet.sh.
- Revert build.gradle.kts, gradle/libs.versions.toml, detekt baseline,
  ModuleManager.kt, and PositionExtrapolation.kt to pre-ML state.
- Keep flake.nix and flake.lock changes (user request).

compileKotlin passes.

superblaubeere27 committed:
Merge latest nextgen (177e1582d) into the invcleaner merge branch

Bring the branch up to date with upstream nextgen (huge divergence from the
April base; latest tip "feat: Wings"). Kept the PR #5938 invcleaner rewrite
authoritative and upgraded the surrounding code onto the merged API.

Conflicts resolved:
- invcleaner/* : keep PR rewrite (ours); dropped remote-only old-style
  GodAxeFacet/SharpAxeFacet.
- ValueGroup.kt : took origin's refactored value-group system and re-added
  the inventoryPreset() DSL (origin removed json()/prepareDeserialize model).
- InventoryPreset.kt / ModuleInventoryCleaner.kt : OffHandSlot ->
  HotbarItemSlot.OFFHAND (origin turned HotbarItemSlot into an enum).
- flake.nix / flake.lock : origin (latest, 26.05).
- GenericSetting.svelte : keep InventoryPresetValue + RegistryMutableListSetting.
- utils.ts : keep portal/clickOutside + isAnniversary.
- Theme SCSS : migrated our inventoryPreset styles from sass $vars to CSS
  custom properties / color-mix (colors.scss now exports only CSS vars).

compileKotlin and theme vite build both pass; svelte-check only reports
nextgen's pre-existing Hud-settings and VirtualList implicit-any errors.

superblaubeere27 committed:
Port GodAxeFacet / SharpAxeFacet onto the merged invcleaner

Restore the sharpness-gated axe specializations that the remote old-style
invcleaner had, adapted to the PR rewrite:

- Re-add GodAxeFacet (axe with Sharpness >= 100, always kept) and
  SharpAxeFacet (axe with Sharpness >= 5, always kept).
- Route axes through getItemFacets: sharpness >= 100 -> GodAxeFacet,
  >= 5 -> SharpAxeFacet, otherwise -> MiningToolItemFacet.

compileKotlin passes.

superblaubeere27 committed:
fix: make nix flake run again

superblaubeere27 committed:
fix: detekt should not care about wildcard imports

superblaubeere27 committed:
fix: use new and improved abstractions

MukjepScarlet committed:
Merge branch 'nextgen' into feat/inventory-manager-such-as-cloudflare-rules

MukjepScarlet committed:
revert stuffs and copyright

MukjepScarlet committed:
[ci skip] reformat

MukjepScarlet committed:
typo

MukjepScarlet committed:
Merge branch 'nextgen' into feat/inventory-manager-such-as-cloudflare-rules

# Conflicts:
#	gradlew.bat
#	src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/player/invcleaner/ItemCategorization.kt
